text/0162-refactoring-support.md

```diff
@@ -0,0 +1,622 @@
+people. What leads you this conclusion? The fact that the person's features
+(face, clothes, height, etc.) are exactly the same in both photographs. But if,
+instead of one person walking, you have two identical siblings, wearing the same
+clothes, you won't be able to tell which one is which, from one photo to the
+next.
+
+The same principle applies here: you can think of stack name + logical ID as a
+place, and the resource properties as its "personal" features. If a resource has
+different properties before and after a potential deployment, the best we can do
+is assume they are different resources. If the properties are the same from one
+point in time to the other, they very likely are the same resource (although the
+developer still has the last word on this). But if there are two resources that
+have the same properties in the same stack, they are like the twin siblings
+case: there's no way to tell them which is which in case they both move to a
+different place.
+
+Indistinguishable resources in this sense are said to be _equivalent_ (see
+Appendix A for a more formal definition). In the unlikely event that two or more
+equivalent resources move, the CLI won't be able to proceed. For example,
+suppose you have two queues in the same stack, named `Queue1` and `Queue2`, with
+the same properties, and without a hard-coded physical ID:
+
+    App
+    └ Stack
+      ├ Queue1
+      └ Queue2
+
+If they get renamed to, let's say, `Queue3` and `Queue4`,
+
+    App
+    └ Stack
+      ├ Queue3
+      └ Queue4
+
+then the CLI will not be able to establish a 1:1 mapping between the old and new
+names. In this case, it will show you the ambiguity, and stop the deployment:
+
+    Ambiguous Resource Name Changes
+    ┌───┬──────────────────────┐
+    │   │ Resource             │
+    ├───┼──────────────────────┤
+    │ - │ Stack.Queue1A4198146 │
+    │   │ Stack.Queue2B0BA5D32 │
+    ├───┼──────────────────────┤
+    │ + │ Stack.Queue3C7606C37 │
+    │   │ Stack.Queue4D681F510 │
+    └───┴──────────────────────┘
+
+    If you want to take advantage of automatic resource refactoring, avoid 
+    renaming or moving multiple identical resources at the same time.
+
+### Refactor only
+
+If you want to execute only the automatic refactoring, use the `cdk refactor`
+command. The behavior is basically the same as with `cdk deploy`: it will detect
+whether there are refactors to be made, ask for confirmation if necessary (
+depending on the flag values), and refactor the stacks involved. But it will
+stop there and not proceed with the deployment. If you only want to see what
+changes would be made, use the `--dry-run` flag.
+
+### Programmatic access
+
+The same refactoring feature is also available in the CDK toolkit library:
+
+```typescript
+declare const toolkit: Toolkit;
+declare const cx: ICloudAssemblySource;
+
+// To execute possible refactors as part of the deploy operation:
+await toolkit.deploy(cx, {
```

Comment:
How does the toolkit library handle the question of equivalent resources? Will it simply fail?

Reply:
Good catch!